lib/pure/ioselectors.nim

```diff
@@ -0,0 +1,2500 @@
+          lock: Lock
+    else:
+      type
+        SelectorImpl[T] = object
+          rSet: TFdSet
+          wSet: TFdSet
+          eSet: TFdSet
+          maxFD: uint
+          fds: Table[SocketHandle, SelectorKey[T]]
+          count: int
+
+    when hasThreadSupport:
+      type Selector*[T] = ptr SelectorImpl[T]
+    else:
+      type Selector*[T] = ref SelectorImpl[T]
+
+    type
+      SelectEventImpl = object
+        rsock: SocketHandle
+        wsock: SocketHandle
+
+    type SelectEvent* = ptr SelectEventImpl
+
+    proc newSelector*[T](): Selector[T] =
+      var maxFD = FD_SETSIZE
+      when hasThreadSupport:
+        result = cast[Selector[T]](allocShared0(sizeof(SelectorImpl[T])))
+        result.maxFD = maxFD.uint
+        result.fds = initSharedTable[SocketHandle, SelectorKey[T]]()
+        initLock result.lock
+      else:
+        result = Selector[T](maxFD: FD_SETSIZE)
+        result.maxFD = maxFD.uint
+        result.fds = initTable[SocketHandle, SelectorKey[T]]()
+
+      FD_ZERO(result.rSet)
+      FD_ZERO(result.wSet)
+      FD_ZERO(result.eSet)
+
+    proc close*(s: Selector) =
+      when hasThreadSupport:
+        deinitSharedTable(s.fds)
+        deallocShared(cast[pointer](s))
+
+    template isEmpty*[T](s: Selector[T]): bool =
+      (s.count == 0)
+
+    template selectAdd[T](s: Selector[T], fd: SocketHandle, event: int) =
+      when hasThreadSupport:
+        withLock s.lock:
+          if (event and EVENT_READ) != 0 and s.rSet.fd_count == FD_SETSIZE:
+            raise newException(ValueError, "Maximum numbers of fds exceeded")
+          if (event and EVENT_WRITE) != 0 and s.wSet.fd_count == FD_SETSIZE:
+            raise newException(ValueError, "Maximum numbers of fds exceeded")
+          if (event and EVENT_READ) != 0:
+            FD_SET(fd, s.rSet)
+          if (event and EVENT_WRITE) != 0:
+            FD_SET(fd, s.wSet)
+            FD_SET(fd, s.eSet)
+          inc(s.count)
+      else:
+        if (event and EVENT_READ) != 0 and s.rSet.fd_count == FD_SETSIZE:
+          raise newException(ValueError, "Maximum numbers of fds exceeded")
+        if (event and EVENT_WRITE) != 0 and s.wSet.fd_count == FD_SETSIZE:
+          raise newException(ValueError, "Maximum numbers of fds exceeded")
+        if (event and EVENT_READ) != 0:
+          FD_SET(fd, s.rSet)
+        if (event and EVENT_WRITE) != 0:
+          FD_SET(fd, s.wSet)
+          FD_SET(fd, s.eSet)
+        inc(s.count)
+
+    proc registerHandle*[T](s: Selector[T], fd: SocketHandle, event: int,
+                         data: T) =
+      var fdi = int(fd)
+      var nkey = SelectorKey[T](ident: fdi, flags: FLAG_HANDLE or event)
+      nkey.key.fd = fdi
+      nkey.key.data = data
+
+      if s.fds.hasKeyOrPut(fd, nkey):
+        raise newException(ValueError, "Re-use of non closed descriptor")
+      selectAdd(s, fd, event)
+
+    proc updateHandle*[T](s: Selector[T], fd: SocketHandle, event: int) =
+      when hasThreadSupport:
```

Comment:
lots of code repetition because of the optional 'withLock()' sections. Guess what. It's a template you can make it a nop when not hasThreadSupport.